### ISSUE TITLE
Shouldn't invalid hashid throw a DoesNotExist instead of a TypeError

### ISSUE THREAD

**Issue by Mactory** [nshafer/django-hashid-field#17]
I am using a hash-id field for one of my models in a small private project. The Hash-ID is used in URLs to access the model instance. I recently had to change the `HASHID_FIELD_SALT` essentially invalidating all existing links, but as it is only a small project, I didn't worry about it too much.

Now some of my users still had old links but as I use `get_object_or_404` to retrieve the model from the database, I thought I'd be save, as I thought the users with the old links would simply get a 404 error. This is, however, not the case. The Hash-ID Field throws a `TypeError: value must be a positive integer or a valid Hashids string` on a old Hash-ID. As the Hash-ID has the same length as the expected Hash-ID, I would expect the Model to throw a `DoesNotExist` Exception that would be caught by `get_object_or_404`.

Is this behavior desired? Otherwise I would propose to change the behavior to throw a `DoesNotExist` if the hash-id is of the expected length but not valid.

I can also provide a pull-request for this issue if you'd consent to my proposal.

**Comment by nshafer** [comment]
Forgive me for a moment while I rubber-duck this... =)

For an AutoField or IntegerField upon which HashidAutoField and HashidField are based on, if you try to look up an ID with something that doesn't cast to an integer, like "asdf" instead of 100 or "100", they will throw a ValueError exception. This is where this behavior originally stems from (though I can't remember why exactly I made it a [TypeError](https://docs.python.org/3/library/exceptions.html#TypeError) instead of [ValueError](https://docs.python.org/3/library/exceptions.html#ValueError)).

Normally this never happens with AutoFields and IntegerFields because people write their URL regexes to only match integers, such as `r'book/(?<pk>[0-9]+)/'`, so doing something like `mysite.com/book/asdf/` instead of `mysite/book/1/` will throw a 404 because Django can't find a valid URL to match "asdf" to. However, if you were to write your URL regexes to accept any character (`[0-9a-zA-Z]`), Django would throw a ValueError exception when the lookup failed. But when we talk about Hashids, it's just not possible to filter out invalid Hashids with a regex as they could possibly be any character, and it's still possible to have an invalid Hashids string made up of valid characters.

(On a side note, with the new URL dispatcher in Django 2.0, I'm thinking I could create a custom Path Converter to allow a path string like `path('book/<hashid:pk>')` that validates the characters and if it's a valid hashid based on the field it's for before it even calls your view. But alas, I'm supporting 1.8-1.11 right now, and even after 2.0 is released, I plan on supporting 1.11 until 2020, same as Django.)

Before version 1.3, the check was done in `get_prep_value()` for both lookups and when you save the instance, and it definitely needs to make sure you're saving either an integer or a valid Hashid string, and throw an exception if not. Just like you can't save, "asdf" to an IntegerField.

However, in the last version (1.3) I created a completely separate code-path for Lookups versus prepping the field to save to the database. I decided for 1.3 to keep the functionality the same (which was throw a TypeError), but I have been giving some thought to changing this now that I have the option. I believe it definitely makes sense to change this for non-exact lookups (iexact, contains, icontains) which would fix #16. However, I'm not exactly sure if I should change it for exact lookups as well.

One thing I've learned, is that the field is used in a lot of really complicated parts of Django that I didn't even know about in the beginning, and I'm afraid there can be some weird side-effects I'm just not even aware of or testing for.

So I have these options:

1. Leave as-is... `filter()`, `get()`, `get_object_or_404()` and other lookup methods will need to be wrapped in try/except blocks to catch the TypeError. This was the only option before 1.3.
2. Change non-exact lookups to an EmptyResultSet (which results in a DoesNotExist down the line), so a lookup like `Book.objects.filter(id__contains="asdf")` will return an empty queryset, but `Book.objects.filter(id="asdf")` will throw a TypeError.
3. Change all lookups to return EmptyResultSet when an invalid Hashids string is given.

And corollary to that, I could add an settings flag to enable throwing exceptions like previously just in case someone was really reliant on that functionality.

What I'm worried about is the weird cases I don't know about, and what this change could mean for them. I personally would like this, as in the project I originally developed the module for, and in the sandbox, I've overridden  `get_object()` to catch the TypeError, but it's messy and ugly, and I would prefer to not have to do it.

```
    def get_object(self, queryset=None):
        try:
            return super(AuthorDetailView, self).get_object(queryset)
        except (ValueError, TypeError):
            raise Http404("Not Found")
```

I don't know... thoughts? Anyone else have any input, especially with cases where they want a lookup to return TypeError instead of just an empty queryset result?

And would option 3 be a breaking change that should warrant a version 2.0? Or is it more of a bug-fix? Should I add an option to enable this EmptyResultSet functionality that is off by default in 1.4, then later after more feedback release 2.0 with the option on by default?

**Comment by nshafer** [comment]
Actually I'm kind of leaning towards implementing option 3, and adding a settings flag to turn on the old behavior just in case someone is relying on it somewhere, as basically a compromise between the two, since this definitely fixes a bug, and I don't know of any cases where you would want a TypeError on lookup.

**Comment by Mactory** [comment]
For my case I would prefer option 3, but I would be just as happy to activate it by a special setting.

I have another thought: I a not cryptographer, but from what one hears, it can sometimes allow for some deductions on the original crypto function or used key/salt if different answers are given in case a key is valid but not in the database or if it is unvalid. Might this be a security weakness if we return different results to the user in the two cases?

If option 3 is implemented, I do believe that it is a breaking change, as it could potentially break a project where the `TypeError` is caught but the `DoesNotExist` isn't. You could think about implementing the change but having to actively enable it with a setting in 1.x. Once you decide to go to 2.0, you could reverse this and make the option the default.

**Comment by nshafer** [comment]
OK, I went with option 3, plus a couple other small changes, and I'm going to call it version 2.0.0 since it has the potential to break existing behavior. Check it out with

```
pip install --upgrade git+https://github.com/nshafer/django-hashid-field.git@empty_lookups
```

I'll take another look at the changes tomorrow, and if nothing else comes to mind, publish it.

**Comment by nshafer** [comment]
This is fixed in 2.0.1. Please read the CHANGELOG for upgrading instructions.
